docs/elasticsearch/ux.md

```diff
@@ -0,0 +1,1240 @@
+
+Elasticsearch allows you to design autocomplete that’s:
+
+- Responsive: Updates suggestions with each keystroke.
+- Relevant: Serves a few but relevant suggestions.
+- Forgiving: Tolerates typos.
+
+Implement autocomplete in one of three ways: prefixes, edge N-grams, and completion suggestors.
+Use these same methods to implement instant search. Instant search is where you show search results as your users type.
+
+### Prefixes (query time)
+
+Use prefix matching to implement autocomplete on any text analyzed field at query time. This is easy because you don’t have to set up special mappings or index your data in a particular way; it simply works with the data that you’ve already indexed.
+
+For example, assume that the user types in the phrase “qui” to search the `text_entry` field. To autocomplete this phrase, use the `match_phrase_prefix` query to search all `text_entry` fields that begin with the prefix "qui".
+To make the word order and relative positions flexible, specify a `slop` value.
+
+```json
+GET shakespeare/_search
+{
+  "query": {
+    "match_phrase_prefix": {
+      "text_entry": {
+        "query": "qui",
+        "slop": 3
+      }
+    }
+  }
+}
+```
+
+Prefix matching is a fairly resource intensive operation. A prefix of `a` could match hundreds of thousands of terms and not be useful to your user.
+To limit the impact of prefix expansion, set `max_expansions` to a reasonable number:
+
+```json
+GET shakespeare/_search
+{
+  "query": {
+    "match_phrase_prefix": {
+      "text_entry": {
+        "query": "qui",
+        "slop": 3,
+        "max_expansions": 10
+      }
+    }
+  }
+}
+```
+
+The ease of implementing query-time autocomplete comes at the cost of performance.
+When implementing this feature on a large scale, we suggest you use an index-time solution instead. With an index-time solution, you might experience slower indexing, but it’s a price you pay only once and not for every query.
+
+### Edge N-grams (index time)
+
+Use edge N-grams to implement autocomplete at index time to improve its performance.
+
+Think of an N-gram as a sliding window on a word, where N stands for the length of the window.
+If we were to N-gram the word "quick", the results would depend on the value of N:
+
+N | Type | N-gram
+:--- | :--- | :---
+1 | Unigram | [ `q`, `u`, `i`, `c`, `k` ]
+2 | Bigram | [ `qu`, `ui`, `ic`, `ck` ]
+3 | Trigram | [ `qui`, `uic`, `ick` ]
+4 | Four-gram | [ `quic`, `uick` ]
+5 | Five-gram | [ `quick` ]
+
+For autocomplete, we only need the beginning N-grams of a search phrase. So, we use a special type of N-gram called edge N-gram.
+
+Edge N-gramming the word "quick" results in:
+
+- `q`
+- `qu`
+- `qui`
+- `quic`
+- `quick`
+
+This follows the sequence we expect the user to type.
+
+Compute the edge N-grams of a text field and match the input queries with it.
```

Comment:
This whole transition feels really disconnected, like notes rather than a paragraph. Stuff to reconsider:

The *reason* edge N-gramming works in a search use case is because it follows the sequence that we expect a user to type.

Why are we instructing users to "compute the edge N-grams?" They don't want to compute anything. They want to tell Elasticsearch to take care of it. I think we can lose this sentence.

"Configuring a field to use edge N-grams is a two-step process. First..."

Reply:
Agreed. Done.